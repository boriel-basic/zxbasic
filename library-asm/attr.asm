; Attribute routines

#include once <sposn.asm>
#include once <error.asm>
#include once <in_screen.asm>
#include once <const.asm>

__ATTR_ADDR:
	; calc start address in DE (as (32 * d) + e)
	; Contributed by Santiago Romero at http://www.speccy.org
	ld h, 0                     ;  7 T-States
	;; ld l, d
	;; add hl, hl   ; HL = HL*2
	;; add hl, hl   ; HL = HL*4
    ld a, d                     ;  4 T-States
    add a, a     ; a * 2        ;  4 T-States
    add a, a     ; a * 4        ;  4 T-States
    ld l, a      ; HL = A * 4   ;  4 T-States

	add hl, hl   ; HL = A * 8   ; 15 T-States
	add hl, hl   ; HL = A * 16  ; 15 T-States
	add hl, hl   ; HL = A * 32  ; 15 T-States
	
	ld d, 18h ; DE = 6144 + E. Note: 6144 is the screen size (before attr zone)
	add hl, de

	ld de, (SCREEN_ADDR)	; Adds the screen address
	add hl, de
	
	; Return current screen address in HL
	ret


; Sets the attribute at a given screen coordinate (D, E).
; The attribute is taken from the ATTR_T memory variable
; Used by PRINT routines
SET_ATTR:

	; Checks for valid coords
	call __IN_SCREEN
	ret nc

__SET_ATTR:
	; Internal __FASTCALL__ Entry used by printing routines
	PROC 

	call __ATTR_ADDR
	ld de, (ATTR_T)	; E = ATTR_T, D = MASK_T

	ld a, d
	and (hl)
	ld c, a	; C = current screen color, masked

	ld a, d
	cpl		; Negate mask
	and e	; Mask current attributes
	or c	; Mix them
	ld (hl), a ; Store result in screen
	
	ret
	
	ENDP


