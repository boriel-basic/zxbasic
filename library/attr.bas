' ----------------------------------------------------------------
' This file is released under the GPL v3 License
' 
' Copyleft (k) 2008
' by Jose Rodriguez-Rosa (a.k.a. Boriel) <http://www.boriel.com>
' ----------------------------------------------------------------

#ifndef __LIBRARY_ATTR__

REM Avoid recursive / multiple inclusion

#define __LIBRARY_ATTR__

#pragma push(case_insensitive)
#pragma case_insensitive = TRUE

' ----------------------------------------------------------------
' function fastcall ATTR
' 
' Parameters: 
'     row: screen row
'     col: screen column
'
' Returns:
'     a byte value containing the screen attribute color value
' ----------------------------------------------------------------
function attr(byval row as ubyte, byval col as ubyte) as ubyte

	' fastcall functions always receive the 1st parameter
	' in accumulator (if byte)
	asm

	PROC
	LOCAL __ATTR_END

    ld e, (ix+7)
    ld d, (ix+5)

    ; Checks for valid coords
    call __IN_SCREEN
    jr nc, __ATTR_END

    call __ATTR_ADDR
    ld a, (hl)	; byte values are returned in accumulator

__ATTR_END:
	ENDP 

	end asm

end function



' ----------------------------------------------------------------
' sub fastcall SETATTR
' 
' Parameters: 
'     row: screen row
'     col: screen column
'	  color: 8bit color attribute
'
' Action: Sets the attribute of screen(row, column) to the given
'		color attribute value.
' ----------------------------------------------------------------
sub setattr(byval row as ubyte, byval col as ubyte, byval value as ubyte)
	' fastcall functions always receive the 1st parameter
	' in accumulator (if byte)
	asm

	PROC
	LOCAL __ATTR_END

    ld e, (ix+7)
    ld d, (ix+5)

    ; Checks for valid coords
    call __IN_SCREEN
    jr nc, __ATTR_END

    call __ATTR_ADDR
	ld a, (ix+9)
    ld (hl), a	; "POKE" attr address, color value)

__ATTR_END:
	ENDP 

	end asm

end sub

#pragma pop(case_insensitive)

' The following is required, because it defines the SCREEN_ADDR constant.
#require "attr.asm"

' The following is required, because it defines the __IN_SCREEN subroutine.
#require "in_screen.asm"

#endif

