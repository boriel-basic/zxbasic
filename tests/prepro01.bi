#include <attr.bas>

PRINT "HOLA"

