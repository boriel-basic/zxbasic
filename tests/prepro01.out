#line 1 "prepro01.bi"
#line 1 "attr.bas"













#pragma push(case_insensitive)
#pragma case_insensitive = TRUE











function attr(byval row as ubyte, byval col as ubyte) as ubyte
    asm

    PROC
    LOCAL __ATTR_END

    ld e, (ix+7)
    ld d, (ix+5)

    
    call __IN_SCREEN
    jr nc, __ATTR_END

    call __ATTR_ADDR
    ld a, (hl)	

__ATTR_END:
    ENDP 

    end asm

end function














sub setattr(byval row as ubyte, byval col as ubyte, byval value as ubyte)
    asm

    PROC
    LOCAL __ATTR_END

    ld e, (ix+7)
    ld d, (ix+5)

    
    call __IN_SCREEN
    jr nc, __ATTR_END

    call __ATTR_ADDR
    ld a, (ix+9)
    ld (hl), a	

__ATTR_END:
    ENDP 

    end asm

end sub











function fastcall attraddr(byval row as ubyte, byval col as ubyte) as uinteger
    asm   
                   
    pop hl         
    ex (sp), hl    
    ld d, a        
    ld e, h
    jp __ATTR_ADDR 
    end asm
end function



#pragma pop(case_insensitive)


#require "attr.asm"


#require "in_screen.asm"

#line 119 "/src/zxb/trunk/library/attr.bas"

#line 2 "prepro01.bi"

PRINT "HOLA"

